subsys/net/lib/sockets/socketpair.c

```diff
@@ -0,0 +1,1087 @@
+	}
+
+	if (spair != NULL && have_local_sem) {
+		k_sem_give(&spair->sem);
+	}
+
+	return res;
+}
+
+static int zsock_poll_update_ctx(struct spair *const spair,
+				 struct zsock_pollfd *const pfd,
+				 struct k_poll_event **pev)
+{
+	int res;
+
+	int signaled;
+	int result;
+
+	struct spair *remote = NULL;
+	bool have_local_sem = false;
+	bool have_remote_sem = false;
+
+	res = k_sem_take(&spair->sem, K_FOREVER);
+	if (res < 0) {
+		NET_ERR("failed to take local sem: %d", res);
+		goto out;
+	}
+
+	have_local_sem = true;
+
+	if (pfd->events & ZSOCK_POLLOUT) {
+		if (sock_is_connected(spair)) {
+
+			remote = z_get_fd_obj(spair->remote,
+				(const struct fd_op_vtable *)
+				&spair_fd_op_vtable, 0);
+
+			__ASSERT(remote != NULL, "remote is NULL");
+
+			res = k_sem_take(&remote->sem, K_FOREVER);
+			if (res < 0) {
+				NET_ERR("failed to take remote sem: %d", res);
+				goto out;
+			}
+
+			have_remote_sem = true;
+
+			if (spair_write_avail(spair) > 0) {
+				pfd->revents |= ZSOCK_POLLOUT;
+			} else {
+				signaled = false;
+				k_poll_signal_check(
+					&remote->read_signal,
+					&signaled, &result);
+				if (signaled) {
+					/* Cannot be SPAIR_SIG_DATA, because
+					 * spair_write_avail() would have
+					 * returned 0
+					 */
+					__ASSERT(result == SPAIR_SIG_CANCEL,
+						"invalid result %d", result);
+					pfd->revents |= ZSOCK_POLLHUP;
+				}
+			}
+		} else {
+			pfd->revents |= POLLHUP;
+		}
+	}
+
+	if (pfd->events & ZSOCK_POLLIN) {
+		if (sock_is_eof(spair)) {
+			pfd->revents |= POLLHUP;
+		} else {
+			if (spair_read_avail(spair) > 0) {
+				pfd->revents |= ZSOCK_POLLIN;
+			} else {
+				signaled = false;
+				k_poll_signal_check(
+					&spair->write_signal,
+					&signaled, &result);
+				if (signaled) {
+					/* Cannot be SPAIR_SIG_DATA, because
+					 * spair_read_avail() would have
+					 * returned 0
+					 */
+					__ASSERT(result == SPAIR_SIG_CANCEL,
+						"invalid result %d", result);
+					pfd->revents |= ZSOCK_POLLIN;
+				}
+			}
+		}
+	}
+
+	(*pev)++;
+
+out:
+	if (remote != NULL && have_remote_sem) {
+		k_sem_give(&remote->sem);
+	}
+
+	if (spair != NULL && have_local_sem) {
+		k_sem_give(&spair->sem);
+	}
+
+	return res;
+}
+
+static int spair_ioctl(void *obj, unsigned int request, va_list args)
+{
+	int res;
+
+	struct zsock_pollfd *pfd;
+	struct k_poll_event **pev;
+	struct k_poll_event *pev_end;
+
+	int flags = 0;
+	struct spair *const spair = (struct spair *)obj;
+
+	if (!spair) {
+		errno = EINVAL;
+		res = -1;
+		goto out;
+	}
+
+	switch (request) {
+		case F_GETFL: {
+			if (sock_is_nonblock(spair)) {
+				flags |= O_NONBLOCK;
+			}
+
+			res = flags;
+			goto out;
+		}
+
+		case F_SETFL: {
+			flags = va_arg(args, int);
+			if (flags & O_NONBLOCK) {
+				spair->flags |= SPAIR_FLAG_NONBLOCK;
+			} else {
+				spair->flags &= ~SPAIR_FLAG_NONBLOCK;
+			}
+
+			res = 0;
+			goto out;
+		}
+
+		case ZFD_IOCTL_CLOSE: {
+			/* disconnect the remote endpoint */
+			spair->remote = -1;
```

Comment:
Should we let remote side know it's alone now?

Reply:
Let's go with yes.

Reply:
I was thinking about that a while ago but ran into race conditions. Good opportunity to address those now.

Reply:
My point that you should have clear understanding how your code behaves when user closes one side of pair, and opens something completely different at the same fd. And it also should be clear to reviewers that you have thought about that. And so far, I didn't see anything explicit ;-).